### DISCUSSION TITLE
Fixed-size buffer of unmanaged Types.

### DISCUSSION THREAD

**Discussion by Qiu233** [dotnet/csharplang#5530] [Language Ideas] [answered]
Recently when performing interops between C++ and C# where a fixed-length array in a struct is needed, I found that keyword `fixed` works, but only for primitive types.
Then I found this spec doc, hoping to find any reason, but only found a snippet of descriptive text as what vs shows:[fixed-size buffer](https://github.com/dotnet/csharplang/blob/main/spec/unsafe-code.md#fixed-size-buffer-declarations)

I used dnspy to show decompiled code and found it just a syntactic sugar.
C# code:
```C#
public unsafe struct FixedStruct
{
	public fixed int A[20];
}
```
```C#
// Token: 0x02000005 RID: 5
public struct FixedStruct
{
	// Token: 0x04000001 RID: 1
	[FixedBuffer(typeof(int), 20)]
	public FixedStruct.<A>e__FixedBuffer A;

	// Token: 0x02000007 RID: 7
	[CompilerGenerated]
	[UnsafeValueType]
	[StructLayout(LayoutKind.Sequential, Size = 80)]
	public struct <A>e__FixedBuffer
	{
		// Token: 0x04000002 RID: 2
		public int FixedElementField;
	}
}
```

As shown above, the nested compiler-constructed struct `<A>e__FixedBuffer` has a fixed `Size` and the actual member `FixedStruct.A` is just a dummy code.
So it's purely a C#  language feature and I see the reason fixed buffers supports only primitives is that the actual size of fixed buffers must be resolved during compilation so that it can be set to `StructLayoutAttribute.Size` of the nested constructed struct.

That's the reason, but I found another fact that `IntPtr`, `UIntPtr` and other pointer Types e.g. `int*` have different sizes on different platforms. In 32bit process, they occupies 4bytes while in 64bit process that's 8bytes, which means `sizeof(T) where T:unmanaged` should not be resolved until type `T` being loaded.

This fact tells that a struct containing fixed buffers could have different size depending on the size of fixed buffers, no matter whether the fixed buffers have variable sizes on different platforms or not, it's all type loading works anyway.

In fact, I thought up a solution to support fixed buffers to all unmanaged types. 
* Just as what csc does now, construct a nested struct invisible to Intellisense.
* Instead of specifying `StructLayoutAttribute.Size`, compiler just fills the the nested struct with members as many as the fixed buffer declaration, regardless of the size of buffer's element.
* The actual member of fixed buffer is set to a declaration of the nested struct, just as csc does now.

Then when the struct gets loaded during run-time, its size will be resolved dynamically with the actual platform info.
In this solution, when compiling, the compiler doesn't need to resolve the actual sizes of buffers, therefore, accepts fixed buffers of any unmanaged type.

Well, there might still be a problem of backwards compatibility.

**Comment by svick** [comment, marked as answer]
I think this is mostly a duplicate of https://github.com/dotnet/csharplang/issues/1314.
